qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,118 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Linear Function."""
+
+from typing import Union, List, Optional
+import numpy as np
+from qiskit.circuit import QuantumCircuit, Gate
+from qiskit.circuit.exceptions import CircuitError
+
+
+class LinearFunction(Gate):
+    """An n_qubit linear circuit."""
+
+    def __init__(
+        self,
+        linear: Union[List[List[int]], np.ndarray, QuantumCircuit],
+        validate_input: Optional[bool] = False,
+    ) -> None:
+        """Create a new linear function.
+
+        Internally, represents a linear function acting on n qubits as a n x n matrix of 0s and 1s.
+
+        Args:
+            linear (list[list] or ndarray or QuantumCircuit):
+                either an n x n matrix, describing the linear function,
+                or a quantum circuit composed of linear gates only
+                (currently supported gates are CX and SWAP).
+
+            validate_input: if True, performs more expensive input validation checks,
+                such as checking that a given n x n matrix is invertible.
+
+        Raises:
+            CircuitError: if the input is invalid:
+                either a matrix is non {square, binary, invertible},
+                or a quantum circuit contains non-linear gates.
+        """
+        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
+            raise CircuitError(
+                "A linear function must be represented either by a list, "
+                "a numpy array, or a quantum circuit with linear gates."
+            )
+
+        if isinstance(linear, QuantumCircuit):
+            # The following function will raise a CircuitError if there are nonlinear gates.
+            self._mat = _linear_quantum_circuit_to_mat(linear)
+
+        else:
+            # Check that the matrix is square
+            if not all(len(row) == len(linear) for row in linear):
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Check that all entries are either 0 or 1
+            for row in linear:
+                for entry in row:
+                    if entry not in (0, 1):
+                        raise CircuitError("A linear function must only contain 0s or 1s.")
+
+            # Optionally, check that the matrix is invertible
+            if validate_input:
+                det = np.linalg.det(linear) % 2
+                if not np.allclose(det, 1):
+                    raise CircuitError(
+                        "A linear function must be represented by an invertible matrix."
+                    )
+
+            self._mat = linear
```

Comment:
You probably want to normalise the type of the input here, so the internal form is always exactly one type, otherwise it's difficult for consumers of the `matrix` to reason about what it will be.  If you normalise it to a Numpy array, you will get most of your checks done significantly faster as well; something like `np.array(linear, dtype=np.bool, copy=False)` will ensure that it's rectangular, and coerce everything to zeros and ones anyway (it's a bit overkill to test that - for large sizes, it will be prohibitively expensive).  To check that it's square it'll just become `len(mat.shape) == 2 and mat.shape[0] == mat.shape[1]`.

Reply:
Thanks! Following your comments, I have normalized the matrix and simplified the checking code.